File: rgctl-tests/ecommerce-javascript/src/coolstore/services/promoService.js
class PromoService {
  constructor() {
    this.percentOffByItem = new Map([['329299', 0.25]]);
  }

  applyCartItemPromotions(shoppingCart) {
    if (!shoppingCart.shoppingCartItemList.length) {
      return;
    }
    for (const sci of shoppingCart.shoppingCartItemList) {
      if (!sci.product) {
        continue;
      }
      const pct = this.percentOffByItem.get(sci.product.itemId);
      if (pct != null) {
        sci.promoSavings = sci.product.price * pct * -1;
        sci.price = sci.product.price * (1 - pct);
      }
    }
  }

  applyShippingPromotions(shoppingCart) {
    if (shoppingCart.cartItemTotal >= 75) {
      shoppingCart.shippingPromoSavings = shoppingCart.shippingTotal * -1;
      shoppingCart.shippingTotal = 0;
    }
  }
}

module.exports = { PromoService };

File: rgctl-tests/ecommerce-cpp/src/coolstore/services/promo_service.cpp
#include "ecommerce/coolstore/services/promo_service.hpp"

namespace ecommerce::coolstore {

PromoService::PromoService() { percentOffByItem_["329299"] = 0.25; }

void PromoService::applyCartItemPromotions(ShoppingCart& shoppingCart) {
    if (shoppingCart.shoppingCartItemList.empty()) return;
    for (auto& sci : shoppingCart.shoppingCartItemList) {
        if (!sci.hasProduct) continue;
        auto it = percentOffByItem_.find(sci.product.itemId);
        if (it != percentOffByItem_.end()) {
            double pct = it->second;
            sci.promoSavings = sci.product.price * pct * -1;
            sci.price = sci.product.price * (1 - pct);
        }
    }
}

void PromoService::applyShippingPromotions(ShoppingCart& shoppingCart) {
    if (shoppingCart.cartItemTotal >= 75) {
        shoppingCart.shippingPromoSavings = shoppingCart.shippingTotal * -1;
        shoppingCart.shippingTotal = 0;
    }
}

}  // namespace ecommerce::coolstore
